### PR TITLE
Creating fat jar with shadow plugin and collisions detector

### PR DESCRIPTION
**Pull request by LepilkinaElena**

### CODE REVIEW

File: tools/kotlin-native-gradle-plugin/build.gradle [hunk 1]
```diff
@@ -77,7 +77,7 @@ dependencies {
     bundleDependencies "org.jetbrains.kotlin:kotlin-native-utils:$kotlinVersion"
     bundleDependencies "org.jetbrains.kotlin:kotlin-native-shared:$konanVersion"
     bundleDependencies "org.jetbrains.kotlin:kotlin-util-io:$kotlinVersion"
-    bundleDependencies "org.jetbrains.kotlin:kotlin-util-klib:$kotlinVersion"
+    //bundleDependencies "org.jetbrains.kotlin:kotlin-util-klib:$kotlinVersion"
```
[the remark this answers is not in the document]
Comment (LepilkinaElena) [reply tools/kotlin-native-gradle-plugin/build.gradle]: As far as I understood, plugin depends on shared that depends on utils-klib. I don't see any problems with build or tests.

**Comment by LepilkinaElena** [comment]
I haven't check yet.

File: tools/kotlin-native-gradle-plugin/build.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (ilmat192) [reply tools/kotlin-native-gradle-plugin/build.gradle]: The `bundleDependencies` configuration includes dependencies that should be included into a plugin fat jar and is not transitive. So all dependencies must be declared here explicitly.

[the remark this answers is not in the document]
Comment (LepilkinaElena) [reply tools/kotlin-native-gradle-plugin/build.gradle]: I already returned this

File: build.gradle [hunk 2]
```diff
@@ -226,42 +249,46 @@ task distSources {
...
+    resolvingRules["META-INF/metadata.kotlin_module"] = "kotlin-compiler"
+    resolvingRules["META-INF/type-system.kotlin_module"] = "kotlin-compiler"
+    resolvingRules["META-INF/util.runtime.kotlin_module"] = "kotlin-compiler"
+    resolvingRules["kotlin/annotation/annotation.kotlin_builtins"] = "kotlin-compiler"
+    resolvingRules["kotlin/collections/collections.kotlin_builtins"] = "kotlin-compiler"
+    resolvingRules["kotlin/coroutines/coroutines.kotlin_builtins"] = "kotlin-compiler"
+    resolvingRules["kotlin/internal/internal.kotlin_builtins"] = "kotlin-compiler"
+    resolvingRules["kotlin/kotlin.kotlin_builtins"] = "kotlin-compiler"
+    resolvingRules["kotlin/ranges/ranges.kotlin_builtins"] = "kotlin-compiler"
+    resolvingRules["kotlin/reflect/reflect.kotlin_builtins"] = "kotlin-compiler"
```
Comment (ilmat192) [inline build.gradle]: Isn't it covered by the regex below?

Comment (ilmat192) [reply build.gradle]: Or the regex should be dropped.

File: build-tools/src/main/kotlin/org/jetbrains/kotlin/CollisionDetector.kt [hunk 3]
```diff
@@ -8,18 +8,29 @@ package org.jetbrains.kotlin
 import org.gradle.api.DefaultTask
 import org.gradle.api.tasks.TaskAction
 import org.gradle.api.artifacts.Configuration
-import org.gradle.api.artifacts.ResolvedArtifact
 import org.gradle.api.tasks.Input
 import org.gradle.api.tasks.InputFiles
 import java.io.File
 
+fun <K, V, T> Map<K, V>.firstOrNull(askedKey: T, predicate:(K, T) -> Boolean): V? {
```
Comment (ilmat192) [inline build-tools/src/main/kotlin/org/jetbrains/kotlin/CollisionDetector.kt]:
```kotlin
inline fun <K, V> Map<K, V>.firstOrNull(predicate: (K, T) -> Boolean): V? 
```

Usage:
```kotlin
resolvingRulesWithRegexes.firstOrNull { it.key.matches(outputPath) }
```

```

File: Interop/Runtime/build.gradle [hunk 4]
```diff
@@ -67,7 +67,7 @@ repositories {
 
 dependencies {
     compile "org.jetbrains.kotlin:kotlin-stdlib:$buildKotlinVersion"
```
Comment (PavelPunegov) [inline Interop/Runtime/build.gradle]: This version of stdlib in dependency seems to be also incorrect.

**Pull request merged**